[pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30143 [pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op**

We would like to integrate the int64 GEMM in FBGEMM into PyTorch. This brings ~4x speedup for the Linear op with LongTensor.

Benchmark code:
```
from __future__ import absolute_import, division, print_function, unicode_literals

import time
import torch

torch.set_num_threads(1)

for M in range(128, 1025, 128):
    N = M
    K = M

    x = torch.LongTensor(M, K)
    w = torch.LongTensor(K, N)

    NITER = 20

    # Test torch.nn.functional.linear
    s = time.time()
    for _ in range(NITER):
        torch.nn.functional.linear(x, w)
        # Z = x @ w
    elapsed_per_iter_linear = (time.time() - s) / NITER

    print(
        "{}, {}, {}, {:0.2f}".format(M, N, K, 2.0 * M * N * K / elapsed_per_iter_linear / 1e9)
    )
```

Before this PR:
```
128, 128, 128, 2.31,
256, 256, 256, 2.49,
384, 384, 384, 2.54,
512, 512, 512, 2.57,
640, 640, 640, 2.46,
768, 768, 768, 2.59,
896, 896, 896, 2.59,
1024, 1024, 1024, 2.61,
```

After this PR:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/int64_gemm]# python torch_linear.py
128, 128, 128, 5.35,
256, 256, 256, 8.34,
384, 384, 384, 9.03,
512, 512, 512, 9.22,
640, 640, 640, 9.55,
768, 768, 768, 9.73,
896, 896, 896, 9.82,
1024, 1024, 1024, 9.63,
```

Differential Revision: [D18610019](https://our.internmc.facebook.com/intern/diff/D18610019/)

jianyuh committed:
[pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op

We would like to integrate the int64 GEMM in FBGEMM into PyTorch. This brings ~4x speedup for the Linear op with LongTensor.

Benchmark code:
```
from __future__ import absolute_import, division, print_function, unicode_literals

import time
import torch

torch.set_num_threads(1)

for M in range(128, 1025, 128):
    N = M
    K = M

    x = torch.LongTensor(M, K)
    w = torch.LongTensor(K, N)

    NITER = 20

    # Test torch.nn.functional.linear
    s = time.time()
    for _ in range(NITER):
        torch.nn.functional.linear(x, w)
        # Z = x @ w
    elapsed_per_iter_linear = (time.time() - s) / NITER

    print(
        "{}, {}, {}, {:0.2f}".format(M, N, K, 2.0 * M * N * K / elapsed_per_iter_linear / 1e9)
    )
```

Before this PR:
```
128, 128, 128, 2.31,
256, 256, 256, 2.49,
384, 384, 384, 2.54,
512, 512, 512, 2.57,
640, 640, 640, 2.46,
768, 768, 768, 2.59,
896, 896, 896, 2.59,
1024, 1024, 1024, 2.61,
```

After this PR:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/int64_gemm]# python torch_linear.py
128, 128, 128, 5.35,
256, 256, 256, 8.34,
384, 384, 384, 9.03,
512, 512, 512, 9.22,
640, 640, 640, 9.55,
768, 768, 768, 9.73,
896, 896, 896, 9.82,
1024, 1024, 1024, 9.63,
```

Differential Revision: [D18610019](https://our.internmc.facebook.com/intern/diff/D18610019/)

[ghstack-poisoned]

nairbv commented on aten/src/TH/generic/THBlas.cpp:
```diff
@@ -380,6 +383,39 @@ void THBlas_(gemm)(
   }
 #endif
+
+#ifdef USE_FBGEMM
```
Since the implementations of these functions will now in some cases use FBGEMM instead of blas/lapack, should we be changing the (file and/or function) names too?

jianyuh replied:
Thanks for the suggestion!

FBGEMM can be viewed as a low precision BLAS (with int8/int64 GEMM as well as other functionality like fp32/int8 SPMDM, etc. for the typical cases in the deep learning). As you can observe here the function signature for `fbgemm::cblas_gemm_i64_i64acc` is quite similar to `cblas_sgemm`.

I am open to change the file/function names. Let me know if you have any good suggestions :)

nairbv replied:
I see the `blas` in the name of the function, so maybe it's fine as is. I'm not really clear on how fbgemm relates to blas, since I was thinking of blas as referring to a specific library. We could consider naming this in another PR if it does make sense.

nairbv commented on aten/src/TH/generic/THBlas.cpp:
```diff
@@ -380,6 +383,39 @@ void THBlas_(gemm)(
 #endif
+
+#ifdef USE_FBGEMM
+#if defined(TH_REAL_IS_LONG)
```
nit: might be cleaner to combine these two, e.g. `#if defined(USE_FBGEMM) && defined(TH_REAL_IS_LONG)` , to reduce nested if defs.

jianyuh replied:
Updated. Thanks!

nairbv commented on aten/src/TH/generic/THBlas.cpp:
```diff
@@ -380,6 +383,39 @@ void THBlas_(gemm)(
+#ifdef USE_FBGEMM
+#if defined(TH_REAL_IS_LONG)
+  if (alpha == 1 && (beta == 0 || beta == 1)) {
+    bool accumulate;
```
I think this can just be `bool accumulate = beta != 0;` or even just in-place could use `!beta`

jianyuh replied:
Updated. Thanks!

nairbv commented on aten/src/TH/generic/THBlas.cpp:
```diff
@@ -380,6 +383,39 @@ void THBlas_(gemm)(
+        transa_ ? fbgemm::matrix_op_t::Transpose
+                : fbgemm::matrix_op_t::NoTranspose,
+        n,
+        m,
```
in https://github.com/pytorch/FBGEMM/blob/f163b30ade066e2e5984ed3c8445ea18825d061b/src/FbgemmI64.cc
I see `M, N, K...` but here we have `n, m, k`. Are we correcting for some difference between naming conventions in  `Blas_(gemm)` vs fbgemm? Maybe that's what the comment above about fortran means but not very clear to me if that's the implication here.

jianyuh replied:
Right: this is related to the comments above. I have added more details in the comments.

Basically we can compute the GEMM with column major C (m x n) = A (m x k) * B (k x n) with the trick of transpose: C^T (n x m) = B^T (n x k) * A^T (k x m), which can be rewritten as the following MACRO: 
```
#define GEMM_COLMAJOR(transf_A, transf_B, m, n, k, alpha, A, lda, B, ldb, beta, C, ldc) \
        GEMM_ROWMAJOR(transf_B, transf_A, n, m, k, alpha, B, ldb, A, lda, beta, C, ldc)
```

nairbv commented on aten/src/TH/generic/THBlas.cpp:
```diff
@@ -380,6 +383,39 @@ void THBlas_(gemm)(
+
+#ifdef USE_FBGEMM
+#if defined(TH_REAL_IS_LONG)
+  if (alpha == 1 && (beta == 0 || beta == 1)) {
```
it looks like beta==1 by default, do we have any unit tests that cover `beta != 1`? I'm not familiar with fbgemm so not sure what happens when `accumulate==false`.

jianyuh replied:
If accumulate== false, we compute C = A * B;
otherwise (accumulate==true), we compute C += A * B.

I think these two cases are mainly used by the Linear module in PyTorch: we rarely use other `beta`.

Re "it looks like beta==1 by default, do we have any unit tests that cover beta != 1":
Right: in "addmm", we need to have beta == 1 as we need to accumulate the the result matrix C. In FBGEMM, we have the test coverage for both accumulate == false/true (beta == 0/1).

nairbv replied:
> we rarely use other beta.

That's what I figured, and even in existing code it's always possible we made a mistake somewhere in how the beta is passed through all these functions. That explanation makes sense though and probably sufficient for this specific change.

jianyuh committed:
Update on "[pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op"


We would like to integrate the int64 GEMM in FBGEMM into PyTorch. This brings ~4x speedup for the Linear op with LongTensor.

Benchmark code:
```
from __future__ import absolute_import, division, print_function, unicode_literals

import time
import torch

torch.set_num_threads(1)

for M in range(128, 1025, 128):
    N = M
    K = M

    x = torch.LongTensor(M, K)
    w = torch.LongTensor(K, N)

    NITER = 20

    # Test torch.nn.functional.linear
    s = time.time()
    for _ in range(NITER):
        torch.nn.functional.linear(x, w)
        # Z = x @ w
    elapsed_per_iter_linear = (time.time() - s) / NITER

    print(
        "{}, {}, {}, {:0.2f}".format(M, N, K, 2.0 * M * N * K / elapsed_per_iter_linear / 1e9)
    )
```

Before this PR:
```
128, 128, 128, 2.31,
256, 256, 256, 2.49,
384, 384, 384, 2.54,
512, 512, 512, 2.57,
640, 640, 640, 2.46,
768, 768, 768, 2.59,
896, 896, 896, 2.59,
1024, 1024, 1024, 2.61,
```

After this PR:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/int64_gemm]# python torch_linear.py
128, 128, 128, 5.35,
256, 256, 256, 8.34,
384, 384, 384, 9.03,
512, 512, 512, 9.22,
640, 640, 640, 9.55,
768, 768, 768, 9.73,
896, 896, 896, 9.82,
1024, 1024, 1024, 9.63,
```

Differential Revision: [D18610019](https://our.internmc.facebook.com/intern/diff/D18610019/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op"


We would like to integrate the int64 GEMM in FBGEMM into PyTorch. This brings ~4x speedup for the Linear op with LongTensor.

Benchmark code:
```
from __future__ import absolute_import, division, print_function, unicode_literals

import time
import torch

torch.set_num_threads(1)

for M in range(128, 1025, 128):
    N = M
    K = M

    x = torch.LongTensor(M, K)
    w = torch.LongTensor(K, N)

    NITER = 20

    # Test torch.nn.functional.linear
    s = time.time()
    for _ in range(NITER):
        torch.nn.functional.linear(x, w)
        # Z = x @ w
    elapsed_per_iter_linear = (time.time() - s) / NITER

    print(
        "{}, {}, {}, {:0.2f}".format(M, N, K, 2.0 * M * N * K / elapsed_per_iter_linear / 1e9)
    )
```

Before this PR:
```
128, 128, 128, 2.31,
256, 256, 256, 2.49,
384, 384, 384, 2.54,
512, 512, 512, 2.57,
640, 640, 640, 2.46,
768, 768, 768, 2.59,
896, 896, 896, 2.59,
1024, 1024, 1024, 2.61,
```

After this PR:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/int64_gemm]# python torch_linear.py
128, 128, 128, 5.35,
256, 256, 256, 8.34,
384, 384, 384, 9.03,
512, 512, 512, 9.22,
640, 640, 640, 9.55,
768, 768, 768, 9.73,
896, 896, 896, 9.82,
1024, 1024, 1024, 9.63,
```

Differential Revision: [D18610019](https://our.internmc.facebook.com/intern/diff/D18610019/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op"


We would like to integrate the int64 GEMM in FBGEMM into PyTorch. This brings ~4x speedup for the Linear op with LongTensor.

Benchmark code:
```
from __future__ import absolute_import, division, print_function, unicode_literals

import time
import torch

torch.set_num_threads(1)

for M in range(128, 1025, 128):
    N = M
    K = M

    x = torch.LongTensor(M, K)
    w = torch.LongTensor(K, N)

    NITER = 20

    # Test torch.nn.functional.linear
    s = time.time()
    for _ in range(NITER):
        torch.nn.functional.linear(x, w)
        # Z = x @ w
    elapsed_per_iter_linear = (time.time() - s) / NITER

    print(
        "{}, {}, {}, {:0.2f}".format(M, N, K, 2.0 * M * N * K / elapsed_per_iter_linear / 1e9)
    )
```

Before this PR:
```
128, 128, 128, 2.31,
256, 256, 256, 2.49,
384, 384, 384, 2.54,
512, 512, 512, 2.57,
640, 640, 640, 2.46,
768, 768, 768, 2.59,
896, 896, 896, 2.59,
1024, 1024, 1024, 2.61,
```

After this PR:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/int64_gemm]# python torch_linear.py
128, 128, 128, 5.35,
256, 256, 256, 8.34,
384, 384, 384, 9.03,
512, 512, 512, 9.22,
640, 640, 640, 9.55,
768, 768, 768, 9.73,
896, 896, 896, 9.82,
1024, 1024, 1024, 9.63,
```

Differential Revision: [D18610019](https://our.internmc.facebook.com/intern/diff/D18610019/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op"


We would like to integrate the int64 GEMM in FBGEMM into PyTorch. This brings ~4x speedup for the Linear op with LongTensor.

Benchmark code:
```
from __future__ import absolute_import, division, print_function, unicode_literals

import time
import torch

torch.set_num_threads(1)

for M in range(128, 1025, 128):
    N = M
    K = M

    x = torch.LongTensor(M, K)
    w = torch.LongTensor(K, N)

    NITER = 20

    # Test torch.nn.functional.linear
    s = time.time()
    for _ in range(NITER):
        torch.nn.functional.linear(x, w)
        # Z = x @ w
    elapsed_per_iter_linear = (time.time() - s) / NITER

    print(
        "{}, {}, {}, {:0.2f}".format(M, N, K, 2.0 * M * N * K / elapsed_per_iter_linear / 1e9)
    )
```

Before this PR:
```
128, 128, 128, 2.31,
256, 256, 256, 2.49,
384, 384, 384, 2.54,
512, 512, 512, 2.57,
640, 640, 640, 2.46,
768, 768, 768, 2.59,
896, 896, 896, 2.59,
1024, 1024, 1024, 2.61,
```

After this PR:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/int64_gemm]# python torch_linear.py
128, 128, 128, 5.35,
256, 256, 256, 8.34,
384, 384, 384, 9.03,
512, 512, 512, 9.22,
640, 640, 640, 9.55,
768, 768, 768, 9.73,
896, 896, 896, 9.82,
1024, 1024, 1024, 9.63,
```

Differential Revision: [D18610019](https://our.internmc.facebook.com/intern/diff/D18610019/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][int64_t gemm][fbgemm] Integrating the int64_t GEMM in FBGEMM into PyTorch Linear op"


We would like to integrate the int64 GEMM in FBGEMM into PyTorch. This brings ~4x speedup for the Linear op with LongTensor.

Benchmark code:
```
from __future__ import absolute_import, division, print_function, unicode_literals

import time
import torch

torch.set_num_threads(1)

for M in range(128, 1025, 128):
    N = M
    K = M

    x = torch.LongTensor(M, K)
    w = torch.LongTensor(K, N)

    NITER = 20

    # Test torch.nn.functional.linear
    s = time.time()
    for _ in range(NITER):
        torch.nn.functional.linear(x, w)
        # Z = x @ w
    elapsed_per_iter_linear = (time.time() - s) / NITER

    print(
        "{}, {}, {}, {:0.2f}".format(M, N, K, 2.0 * M * N * K / elapsed_per_iter_linear / 1e9)
    )
```

Before this PR:
```
128, 128, 128, 2.31,
256, 256, 256, 2.49,
384, 384, 384, 2.54,
512, 512, 512, 2.57,
640, 640, 640, 2.46,
768, 768, 768, 2.59,
896, 896, 896, 2.59,
1024, 1024, 1024, 2.61,
```

After this PR:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/int64_gemm]# python torch_linear.py
128, 128, 128, 5.35,
256, 256, 256, 8.34,
384, 384, 384, 9.03,
512, 512, 512, 9.22,
640, 640, 640, 9.55,
768, 768, 768, 9.73,
896, 896, 896, 9.82,
1024, 1024, 1024, 9.63,
```

Differential Revision: [D18610019](https://our.internmc.facebook.com/intern/diff/D18610019/)

[ghstack-poisoned]
